contracts/pob/assembly/Pob.ts

```diff
@@ -118,40 +148,42 @@ export class Pob {
     System.require(args.header!.timestamp % params.quantum_length == 0, "time stamp does not match time quanta");
 
     // Get signer's public key
-    const registration = System.getObject<Uint8Array, pob.public_key_record>(State.Space.REGISTRATION, signer, pob.public_key_record.decode);
+    const registration = System.getObject<Uint8Array, pob.public_key_record>(State.Space.Registration(), args.header!.signer!, pob.public_key_record.decode);
     System.require(registration != null, "signer address has no public key record");
 
     // Create vrf payload and serialize it
-    const payload = new pob.vrf_payload(metadata.seed, timestamp);
+    const payload = new pob.vrf_payload(metadata.seed, args.header!.timestamp);
     const payload_bytes = Protobuf.encode(payload, pob.vrf_payload.encode);
 
     System.require(System.verifyVRFProof(registration!.public_key!, signature.vrf_proof!, signature.vrf_hash!, payload_bytes), "proof failed vrf validation");
 
     // Ensure vrf hash divided by producer's vhp is below difficulty
-    const difficulty = BigIntFromBytes(metadata.difficulty!);
-    const target = Constants.U256_MAX.div(difficulty);
+    const difficulty = u128.fromBytes(metadata.difficulty!, true);
+    const target = u128.Max / difficulty;
     let mh = new Crypto.Multihash();
     mh.deserialize(signature.vrf_hash!)
-    const hash = BigIntFromBytes(mh.digest);
-    const vhp_balance = BigInt.from(vhp.balanceOf(signer));
+    // This purposefully only gets the upper 128 bits. That's all we need
+    let hash = u128.fromBytes(mh.digest, true);
+    const vhp_balance = u128.fromU64(vhp.balanceOf(args.header!.signer!));
 
-    const mark = hash.div(vhp_balance);
+    hash = hash / vhp_balance;
 
-    System.require(mark < target, "provided hash is not sufficient");
+    System.require(hash < target, "provided hash is not sufficient");
 
-    const virtual_supply = koin.totalSupply() + vhp.totalSupply();
-    const yearly_inflation = virtual_supply * params.target_annual_inflation_rate / Constants.FIXED_POINT_PRECISION;
+    const virtual_supply = u128.fromU64(koin.totalSupply() + vhp.totalSupply());
```

Comment:
This looks like it might overflow when total supplies reach near max u64.

Reply:
Yes, it could. That's bad for everything. But I ran the numbers, and at an inflation rate of 2%, we won't overflow the virtual supply for 542 years. I can make the change here, but overflowing the virtual supply has bad ramifications for far more than just this calculation. If the burn percent changes too much in one direction, it means we could overflow the koin supply or the vhp supply. I actually think leaving the overflow here makes the most sense because it would overflow addition to somewhere near 0 and practically stop minting of new KOIN. That is in some sort of run away scenario caused by some other bug. In all likelihood, we'll have some other solution figured out within the next 500 years.